### PR TITLE
Enhancement for exclusive revamp

### PR DESCRIPTION
**Pull request by aashishAil**
## Relevant Documentation Links
- https://www.notion.so/creatorstack/Epic-Redesigning-Digital-Download-Exclusive-44c6367cdce84cefb48eca22f864f28d

## Relevant Jira Links
- https://creatorstack.atlassian.net/browse/BCS-1327

## Other Relevant PRs
- None

## Changes
- Added generic cache instance
- Integrated Redis instance
- Added mocks for cache instance
- Added new post type for exclusive
- Added number util

### Changes Include

-   [ ] Bug fix 🐛 (non-breaking change which fixes an issue)
-   [ ] New feature ✨ (non-breaking change which adds functionality)
-   [ ] Breaking change 😢 (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Refactor ⚒️ (code change that neither fixes a bug nor adds a feature)
-   [ ] Tooling ⚙️ (Changes to the build process or auxiliary tools and libraries such as documentation generation)
-   [ ] Performance ⚡ (code change that improves performance)

### Checklist

-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [ ] My changes generate no new warnings
-   [ ] Requires migrations
-   [ ] Requires dependency update
-   [ ] Breaking API Change
-   [ ] I have added proper logs for production debugging
-   [ ] I have added performance (APM) monitoring checkpoints
-   [ ] I have added unit tests

## Screenshots (if any)
(prefer animated gif)

### Test #1

Description

- [ ] Acceptance Creteria #1
- [ ] Acceptance Creteria #2

### CODE REVIEW

File: database/cache.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	SetValue(ctx context.Context, key string, value interface{}, expiryInSeconds *time.Duration) error
+	GetValue(ctx context.Context, key string) (interface{}, error)
+}
+
+type cacheInstance struct {
+	client         *redis.Client
+	instanceConfig CacheConfig
+}
+
+func (d *cacheInstance) initialize() {
```
Comment (MrDHat) [inline database/cache.go:20]: Can we inject the client instead? It'll help us with mocks

Comment (MrDHat) [reply database/cache.go:20]: Ideally it should work just like postgres. Something like this: https://github.com/CreatorStack/backbone/blob/728c5cba1f1af9ce1cd76f5e4c339c4818a550a9/database/redis.go

File: database/cache.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	value, err := d.client.Get(ctx, key).Result()
+	if err != nil {
+		logger.Log.WithError(err).Error(groupError)
+		return nil, err
+	}
+
+	return value, nil
+}
+
+func NewCacheInstance(instanceConfig CacheConfig) CacheInstance {
```
Comment (MrDHat) [inline database/cache.go:64]: Also, we should move it to a package called `cache` inside backbone

File: utils/number.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	} else if (a == nil && b != nil) || (a != nil && b == nil) {
+		return true
+	} else if *a != *b {
+		return true
+	}
+
+	return false
+}
+
+func NewNumber() Number {
```
Comment (MrDHat) [inline utils/number.go:162]: do we really need injection here? This is a generic utility & injection seems overdoing it here

**Comment by MrDHat** [comment]
failing deepsource should be fixed before merge

File: utils/number.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
+package utils
+
+type Number interface {
+	CheckNullAndSetInt(a, b *int) *int
```
[the remark this answers is not in the document]
Comment (krishnamurthypranesh) [reply utils/number.go:4]: Generics are available in 1.18, not 1.17

**Review by krishnamurthypranesh — APPROVED** [review]

File: database/cache.go (hunk 2 above)
Comment (aashishAil) [reply database/cache.go]: Done

File: utils/number.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aashishAil) [reply utils/number.go:4]: https://creatorstack.atlassian.net/browse/BCS-1504

File: utils/number.go (hunk 3 above)
Comment (aashishAil) [reply utils/number.go:162]: As discussed, it makes testing easier

**Review by MrDHat — APPROVED** [review]

**Pull request merged**